src/esp/gfx/CubeMapCamera.cpp

```diff
@@ -0,0 +1,121 @@
+  updateOriginalViewingMatrix();
+}
+CubeMapCamera::CubeMapCamera(scene::SceneNode& node,
+                             const vec3f& eye,
+                             const vec3f& target,
+                             const vec3f& up)
+    : CubeMapCamera(node,
+                    Mn::Vector3{eye},
+                    Mn::Vector3{target},
+                    Mn::Vector3{up}) {}
+CubeMapCamera::CubeMapCamera(scene::SceneNode& node,
+                             const Mn::Vector3& eye,
+                             const Mn::Vector3& target,
+                             const Mn::Vector3& up)
+    : RenderCamera(node, eye, target, up) {
+  updateOriginalViewingMatrix();
+}
+CubeMapCamera& CubeMapCamera::updateOriginalViewingMatrix() {
+  originalViewingMatrix_ = this->node().transformation();
+  return *this;
+}
+
+CubeMapCamera& CubeMapCamera::restoreTransformation() {
+  this->node().setTransformation(originalViewingMatrix_);
+  return *this;
+}
+
+CubeMapCamera& CubeMapCamera::switchToFace(unsigned int cubeSideIndex) {
+  switch (cubeSideIndex) {
+    case 0:
+      switchToFace(Mn::GL::CubeMapCoordinate::PositiveX);
+      break;
+    case 1:
+      switchToFace(Mn::GL::CubeMapCoordinate::NegativeX);
+      break;
+    case 2:
+      switchToFace(Mn::GL::CubeMapCoordinate::PositiveY);
+      break;
+    case 3:
+      switchToFace(Mn::GL::CubeMapCoordinate::NegativeY);
+      break;
+    case 4:
+      switchToFace(Mn::GL::CubeMapCoordinate::PositiveZ);
+      break;
+    case 5:
+      switchToFace(Mn::GL::CubeMapCoordinate::NegativeZ);
+      break;
+    default:
+      LOG(ERROR)
+          << "Warning: CubeMapCamera::switchToFace(): the index of cube side "
+          << cubeSideIndex
+          << " is illegal. Camera orientation stays unchanged.";
+      break;
+  }
+  return *this;
+}
+
+CubeMapCamera& CubeMapCamera::switchToFace(Mn::GL::CubeMapCoordinate cubeSide) {
+  Mn::Vector3 eye{0.0, 0.0, 0.0};
+  Mn::Vector3 yUp{0.0, 1.0, 0.0};
+  Mn::Vector3 zUp{0.0, 0.0, 1.0};
+
+  auto cameraLocalTransform = [&]() {
+    switch (cubeSide) {
+      case Mn::GL::CubeMapCoordinate::PositiveX:
+        return Mn::Matrix4::lookAt(eye, Mn::Vector3{-1.0, 0.0, 0.0}, -yUp);
+        break;
+      case Mn::GL::CubeMapCoordinate::NegativeX:
+        return Mn::Matrix4::lookAt(eye, Mn::Vector3{1.0, 0.0, 0.0}, -yUp);
+        break;
+      case Mn::GL::CubeMapCoordinate::PositiveY:
+        return Mn::Matrix4::lookAt(eye, Mn::Vector3{0.0, 1.0, 0.0}, -zUp);
+        break;
+      case Mn::GL::CubeMapCoordinate::NegativeY:
+        return Mn::Matrix4::lookAt(eye, Mn::Vector3{0.0, -1.0, 0.0}, zUp);
+        break;
+      case Mn::GL::CubeMapCoordinate::PositiveZ:
+        return Mn::Matrix4::lookAt(eye, Mn::Vector3{0.0, 0.0, -1.0}, -yUp);
+        break;
+      case Mn::GL::CubeMapCoordinate::NegativeZ:
+        return Mn::Matrix4::lookAt(eye, Mn::Vector3{0.0, 0.0, 1.0}, -yUp);
+        break;
```

Comment:
The `break`s are not needed after a return -- some compilers will even warn about those being unreachable.

Reply:
Facebook coding convention requires "Always end a non-empty case with a break." (see here: [the public link](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es78-dont-rely-on-implicit-fallthrough-in-switch-statements)). Of course I understand here it is a special case because it is a "return" before the "break". But I prefer to keep the breaks here. It is a good habit and less error-prone in case any refactoring happen in the future.

So far I did not see any compiler give me such warnings, neither on MacOS nor on Linux. Thanks for the heads up!

Reply:
> So far I did not see any compiler give me such warnings

Not sure if `-Wpedantic -Wextra` is enabled on Habitat code, but I remember either those or the `/W4` warning level on MSVC complaining about "unreachable code after return". Or I might be misremembering, or maybe this particular case isn't checked. Conversely, on recent compilers these warning levels yell at you when you forget an actually important `break`, and I don't think they would warn here :)

> Always end a non-empty case with a break

Every ~~exception has a rule~~ rule has an exception, from my PoV the `break` is never reached and thus extraneous.

Reply:
>Conversely, on recent compilers these warning levels yell at you when you forget an actually important break

It is quite dangerous. What if the developer ignores such warnings.

Yes, I totally understand your points here. It can reduce the noise from certain compilers, and make the code a bit simpler.

However, like I said, I believe it is a good habit to have a `break` in each case, and it follows Facebook coding convention we agreed on. It is just like we require `{}` in each if clause no matter if it is just one line of code.

More importantly, I did not see any significant benefits from making an exception to the rule. It just make one or two specific compiler less noisy (even the noise is not harmful). If we remove the `break`, however, it makes the code potentially more error-prone in the future refactoring. 

Shall we agree on the disagree here? Hope it is not deal breaker in order to get your approval on the code. :)

Reply:
Yeah it's not a deal breaker for me, just sharing my view :)

> What if the developer ignores such warnings.

`-Werror=implicit-fallthrough` turns that into a *loud yell* :P 

And (just FYI) for a portable `[[fallthrough]]` there's a [CORRADE_FALLTHROUGH](https://doc.magnum.graphics/corrade/Macros_8h.html#af3cd136001e4eb1fc1a78aa17fe8232e).